### DISCUSSION TITLE
Support other databases for music library.

### DISCUSSION THREAD

**Discussion by konikvranik** [music-assistant/support#3353] [Feature requests and ideas]
It looks like MA uses SQLite to store library data, which is extremely slow on my setup with NFS.
It would be nice to allow MA to also use other engines like Postgres and so on.
I saw a previous request from 2022 that is locked and not implemented, but IMHO it's back on the table again.

**Comment by konikvranik** [comment]
Actually, it blocks me from using MA. I like the project; I'd like to support it, but I cannot use it because the search is so slow, so it's unusable with SQLite.

**Reply by marcelveldt** [reply]
can you try it again with 2.4.0 beta 15 ?

We've been working with somebody with a super large library (over 500.000 local files) and the performance is now solid.
That is, the db is still on local storage, not on remote NFS but as long as your link is fast that should not matter that much.

Can you give it a try ?

Also do note that when you are using the global search and you have streaming providers attached, those will be queried as well and those can be slow. For indication of performance, do a search on one of the library pages, such as artists, albums etc.

Switching to another db engine is very much likely not going to happen any time soon.

**Reply by konikvranik** [reply]
Thank you for your remark. I'm going to try with 2.4.0b15.

Regarding providers, I have Tidal, YT Music, TuneIn radio and 3 Samba shares. Looks like local Samba providers are the performance killers:
Without it, it's fast enough. As soon as I add these Samba shares, its indexing them for several hours. There are mostly audiobooks. From the start, search is still acceptable, but as soon as it indexes it, search is almost unusable. A single query takes several minutes and returns irrelevant results (mostly previous query).
With 2.4.0b14 the search is faster and mostly returning relevant results, but still slow after reindexing. Let's see how it's going to behave on b15.

**Reply by konikvranik** [reply]
BTW, during indexing it sometimes throws:
```
2025-01-16 12:56:47.888 ERROR (MainThread) [music_assistant.metadata] Error while updating metadata for library://artist/276: artist not found in library: 276
Traceback (most recent call last):
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/metadata.py", line 750, in _process_metadata_lookup_jobs
    item = await self.mass.music.get_item_by_uri(item_uri)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/music.py", line 491, in get_item_by_uri
    return await self.get_item(
           ^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/music.py", line 515, in get_item
    return await ctrl.get(
           ^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/media/base.py", line 284, in get
    if library_item := await self.get_library_item_by_prov_id(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/media/base.py", line 400, in get_library_item_by_prov_id
    return await self.get_library_item(item_id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/venv/lib/python3.12/site-packages/music_assistant/controllers/media/base.py", line 389, in get_library_item
    raise MediaNotFoundError(msg)
music_assistant_models.errors.MediaNotFoundError: artist not found in library: 276

```
for different IDs of artists

**Reply by OzGav** [reply]
How long ago did this DB get built from scratch? Having missing artist DB entries isn't normal

**Reply by konikvranik** [reply]
After upgrade to b15 I cleared cache, reset DB and let Samba providers to refresh. So it happened right after DB reset.

**Reply by konikvranik** [reply]
So, it took ~ 4 and 3/4 hours to reindex Samba.

There are 289 artists in the library, 304 albums, 3937 tracks, 2 playlists, and 8 radios.

version: 2.4.0b15

Search is slow still.

**Reply by konikvranik** [reply]
If I try searching on the artist's page for David Bowie, it takes some time and then returns no results, recommending to use global search.
If the DB is empty and I use global search, results are returned immediately from Tidal and YT Music.

**Reply by konikvranik** [reply]
The speed of NFS is 4,1 MB/s on write and 3,7 MB/s on read (tested by dd). The path between RPi and NFS is on a wired network. MA runs on RPi5 with either 4 or 8 GB of RAM.

**Reply by OzGav** [reply]
Don’t know what to tell you I have a slightly larger library and search works fine. Maybe try with just one share and see if that makes a difference but I’m grasping at straws.

**Reply by resnostyle** [reply]
I know its a big ask, but I'd like support for other database engines aside from sqlite. I'm looking to using kube to run this, and what ever I can get into a database server I'm all for. Thanks for a great software solution though.

**Reply by Jc2k** [reply]
@resnostyle I run it on a 4 node k8s cluster with the SQLite db on longhorn storage. Cluster is a hodge podge of under specced eBay specials, mostly mini micro nodes. 817 artists. 383 albums. 4978 tracks. Performance is fine here.

(K8s isn't really supported anyway, the networking will kill you before storage).

**Reply by marcelveldt** [reply]
> The speed of NFS is 4,1 MB/s on write and 3,7 MB/s on read (tested by dd). The path between RPi and NFS is on a wired network. MA runs on RPi5 with either 4 or 8 GB of RAM.

That is a bit slow tbh. Fine for the fileprovider (the sync will just be slow) but you shouldnt be hosting the database on that. Just use local storage for the system and store the music on some remote location.

**Comment by marcelveldt** [comment]
Folks, there is little to no chance we are going to switch our database (which is a 6 months job alone) just for a few edge case setups. We have some recommended setups for which we have the manpower to provide support and those run fine, even for super large collections. We tweaked the db and cache and we keep doing that when we see there's a need. sqlite is just fine for what we need. 

So feel free to offer help in improving the current sqlite experience, queries etc. but switching to another db engine is not going to happen unless there is a real pressing need, which I think there is not at the moment.
